### PR TITLE
TF-1604: Add requiredCapabilities support TeamMailbox for mailbox

### PR DESCRIPTION
**Pull request by nqhhdev**
- Issue: #1604  

- Add requiredCapabilities support TeamMailbox for mailbox
- Add requiredCapabilities support TeamMailbox for `SpamReport`
- Add requiredCapabilities support TeamMailbox for `GetEmailMethod`
- Add method check and get capability is support `TeamMailboxes`
- Add requiredCapabilities support TeamMailbox for `SetEmailMethod`

`DEMO`

https://user-images.githubusercontent.com/99852347/227088690-ad1874e7-1ae1-4a52-98d0-c1962c6077b6.mov

### CODE REVIEW

**Review by hoangdat — COMMENTED** [review]
```
 Set<CapabilityIdentifier> _capabilitiesForMailboxMethod(Session session, AccountId accountId) {
    final getMailboxCreated = GetMailboxMethod(accountId);
    try {
     requireCapability(
        session,
        accountId,
        [CapabilityIdentifier.jmapTeamMailboxes]);
      return getMailboxCreated.requiredCapabilitiesSupportTeamMailboxes;
    } catch (_) {
      return getMailboxCreated.requiredCapabilities;
    }
  }
```
please refactor this method also, why we need `final getMailboxCreated = GetMailboxMethod(accountId);`

File: lib/features/base/base_controller.dart [hunk 1]
```diff
@@ -237,7 +237,7 @@ abstract class BaseController extends GetxController
           await LocalNotificationManager.instance.setUp();
         }
         FcmInteractorBindings().dependencies();
-        FcmController.instance.initialize(accountId: accountId);
+        FcmController.instance.initialize(session: session,accountId: accountId);
```
Comment (hoangdat) [inline lib/features/base/base_controller.dart:240]:
```suggestion
        FcmController.instance.initialize(session: session, accountId: accountId);
```

File: lib/features/email/data/network/email_api.dart [hunk 2]
```diff
@@ -63,7 +64,24 @@ class EmailAPI with HandleSetErrorMixin {
 
   EmailAPI(this._httpClient, this._downloadManager, this._dioClient);
 
-  Future<Email> getEmailContent(AccountId accountId, EmailId emailId) async {
+  Set<CapabilityIdentifier> _capabilitiesForEmailMethod(Session session, AccountId accountId) {
+    final getMailboxCreated = GetEmailMethod(accountId);
```
Comment (hoangdat) [inline lib/features/email/data/network/email_api.dart:68]: this does not make sense to me. How about: 
```
Set<CapabilityIdentifier> _capabilitiesForSupportedTeamMailboxes(Session session, AccountId accountId, Set<CapabilityIdentifier> requiredCapabilities) {
  //....
}
```

**Comment by nqhhdev** [comment]
> ```
>  Set<CapabilityIdentifier> _capabilitiesForMailboxMethod(Session session, AccountId accountId) {
>     final getMailboxCreated = GetMailboxMethod(accountId);
>     try {
>      requireCapability(
>         session,
>         accountId,
>         [CapabilityIdentifier.jmapTeamMailboxes]);
>       return getMailboxCreated.requiredCapabilitiesSupportTeamMailboxes;
>     } catch (_) {
>       return getMailboxCreated.requiredCapabilities;
>     }
>   }
> ```
> 
> please refactor this method also, why we need `final getMailboxCreated = GetMailboxMethod(accountId);`

I use it to call `requiredCapabilitiesSupportTeamMailboxes` and `requiredCapabilities` available in `jmap-dart-client`. 
If it's not clear then maybe we'll fix it. WDYT  let me know.

File: lib/features/thread/data/network/thread_api.dart [hunk 3]
```diff
@@ -17,14 +19,33 @@ import 'package:jmap_dart_client/jmap/mail/email/get/get_email_response.dart';
...
+          [CapabilityIdentifier.jmapTeamMailboxes]);
+      return {
+        CapabilityIdentifier.jmapCore,
+        CapabilityIdentifier.jmapMail,
+        CapabilityIdentifier.jmapTeamMailboxes
+      };
+    } catch (_) {
+      return getMailboxCreated.requiredCapabilities;
+    }
+  }
```
Comment (hoangdat) [inline lib/features/thread/data/network/thread_api.dart:45]: move to `Capability_validator`. Please dont' use `final getMailboxCreated = GetEmailMethod(accountId);` in this method

File: lib/main/error/capability_validator.dart [hunk 4]
```diff
@@ -28,6 +28,17 @@ void requireCapability(Session session, AccountId accountId, List<CapabilityIden
 
 extension CapabilityIdentifierExtension on List<CapabilityIdentifier> {
 
+  List<CapabilityIdentifier> capabilitiesSupportTeamMailboxes(Session session, AccountId accountId, List<CapabilityIdentifier> capabilities) {
```
Comment (dab246) [inline lib/main/error/capability_validator.dart:31]:
```suggestion
 List<CapabilityIdentifier> toCapabilitiesSupportTeamMailboxes(Session session, AccountId accountId) {
    try {
      requireCapability(session, accountId, [CapabilityIdentifier.jmapTeamMailboxes]);
      add(CapabilityIdentifier.jmapTeamMailboxes);
     return this;
    } catch (error) {
      logError('CapabilityIdentifierExtension::toCapabilitiesSupportTeamMailboxes(): $error');
      return this;
    }
  }
```

File: lib/l10n/intl_messages.arb [hunk 5]
```diff
@@ -1,5 +1,5 @@
 {
-  "@@last_modified": "2023-03-10T14:00:31.546227",
+  "@@last_modified": "2023-03-22T15:59:57.628888",
```
Comment (dab246) [inline lib/l10n/intl_messages.arb:2]: Don't change it if not add text

File: lib/features/mailbox_dashboard/data/network/spam_report_api.dart [hunk 6]
```diff
@@ -35,8 +38,12 @@ class SpamReportApi {
           ReferencePath.idsPath,
         ));
     final getMailboxInvocation = requestBuilder.invocation(getMailBoxMethod);
+
+    final capabilities = getMailBoxMethod.requiredCapabilities.toList()
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/data/network/spam_report_api.dart:42]: why we need to convert `toList` then `toSet`? we can do better.

Comment (dab246) [reply lib/features/mailbox_dashboard/data/network/spam_report_api.dart:42]: U can write an `extension` for Set<Capability>

File: lib/main/error/capability_validator.dart [hunk 7]
```diff
@@ -37,4 +37,18 @@ extension CapabilityIdentifierExtension on List<CapabilityIdentifier> {
...
   }
+}
+
+extension CapabilityIdentifierSetExtension on Set<CapabilityIdentifier> {
+
+  Set<CapabilityIdentifier> toCapabilitiesSupportTeamMailboxes(Session session, AccountId accountId) {
+    try {
+      requireCapability(session, accountId, toList());
+      add(CapabilityIdentifier.jmapTeamMailboxes);
+      return toSet();
```
Comment (dab246) [inline lib/main/error/capability_validator.dart:48]:
```suggestion
return this;
```

File: lib/main/error/capability_validator.dart [hunk 8]
```diff
@@ -37,4 +37,18 @@ extension CapabilityIdentifierExtension on List<CapabilityIdentifier> {
...
+    } catch (error) {
+      logError('CapabilityIdentifierExtension::isSupported(): $error');
+      return toSet();
```
Comment (dab246) [inline lib/main/error/capability_validator.dart:51]: same

File: lib/main/error/capability_validator.dart [hunk 9]
```diff
@@ -37,4 +37,18 @@ extension CapabilityIdentifierExtension on List<CapabilityIdentifier> {
       return false;
     }
   }
+}
+
+extension CapabilityIdentifierSetExtension on Set<CapabilityIdentifier> {
+
+  Set<CapabilityIdentifier> toCapabilitiesSupportTeamMailboxes(Session session, AccountId accountId) {
+    try {
+      requireCapability(session, accountId, toList());
```
Comment (dab246) [inline lib/main/error/capability_validator.dart:46]: Call `requireCapability` for capability support `Team-Mailbox`
```suggestion
requireCapability(session, accountId, [CapabilityIdentifier.jmapTeamMailboxes]);
```

**Review by dab246 — APPROVED** [review]

**Review by hoangdat — APPROVED** [review]

**Pull request merged**